File: 49_group_anagrams/group_anagrams.go
package main

import (
	"strconv"
	"strings"
)

type Solution struct{}

func (s *Solution) GroupAnagrams(strs []string) [][]string {
	groups := make(map[string][]string)

	for _, word := range strs {
		count := [26]int{}
		for _, ch := range word {
			count[ch-'a']++
		}

		keyParts := make([]string, 26)
		for i, v := range count {
			keyParts[i] = strconv.Itoa(v)
		}
		key := strings.Join(keyParts, "#")

		groups[key] = append(groups[key], word)
	}

	result := make([][]string, 0, len(groups))
	for _, group := range groups {
		result = append(result, group)
	}
	return result
}

func main() {
	solution := Solution{}
	words := []string{"eat", "tea", "tan", "ate", "nat", "bat"}
	result := solution.GroupAnagrams(words)

	for _, group := range result {
		println(strings.Join(group, ", "))
	}
}

File: 49_group_anagrams/group_anagrams.py
from typing import List
from collections import defaultdict

class Solution:
    def group_anagrams(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)

        for word in strs:
            count = [0] * 26
            for ch in word:
                count[ord(ch) - ord('a')] += 1
            
            key = tuple(count)
            groups[key].append(word)

        return list(groups.values())


if __name__ == "__main__":
    solution = Solution()
    words = ["eat", "tea", "tan", "ate", "nat", "bat"]
    result = solution.group_anagrams(words)

    for v in result: print(v)
